lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,923 @@
+                if version != req_version:
+                    break
+        else:
+            return True
+
+        # The loop was broken early, it does not meet all the requirements
+        return False
+
+    @staticmethod
+    def from_tar(path, validate_certs, force, parent=None):
+        if not tarfile.is_tarfile(path):
+            raise AnsibleError("Collection artifact at '%s' is not a valid tar file." % to_native(path))
+
+        info = {}
+        with tarfile.open(path, mode='r') as collection_tar:
+            for member_name, property_name in CollectionRequirement._FILE_MAPPING:
+                try:
+                    member = collection_tar.getmember(member_name)
+                except KeyError:
+                    raise AnsibleError("Collection at '%s' does not contain the required file %s."
+                                       % (to_native(path), to_native(member_name)))
+
+                with _tarfile_extract(collection_tar, member) as member_obj:
+                    try:
+                        info[property_name] = json.loads(to_text(member_obj.read()))
+                    except ValueError:
+                        raise AnsibleError("Collection tar file %s is not a valid json string." % to_native(member_name))
+
+        meta = info['manifest']['collection_info']
+        files = info['files']['files']
+
+        namespace = meta['namespace']
+        name = meta['name']
+        version = meta['version']
+
+        return CollectionRequirement(namespace, name, path, None, [version], version, force, parent=parent,
+                                     validate_certs=validate_certs, metadata=meta, files=files)
+
+    @staticmethod
+    def from_path(path, validate_certs, force, parent=None):
+        info = {}
+        for file_name, property_name in CollectionRequirement._FILE_MAPPING:
+            file_path = os.path.join(path, file_name)
+            if os.path.exists(file_path):
+                with open(file_path, 'r') as file_obj:
+                    try:
+                        info[property_name] = json.load(file_obj)
+                    except ValueError:
+                        raise AnsibleError("Collection file at '%s' is not a valid json string." % file_path)
+
+        if 'manifest' in info:
+            meta = info['manifest']['collection_info']
+        else:
+            display.warning("Collection at '%s' does not have a MANIFEST.json file, cannot detect version."
+                            % to_text(path))
+            parent_dir, name = os.path.split(path)
+            namespace = os.path.split(parent_dir)[1]
+            meta = {
+                'namespace': namespace,
+                'name': name,
+                'version': '*',
+                'dependencies': {},
+            }
+
+        namespace = meta['namespace']
+        name = meta['name']
+        version = meta['version']
+
+        files = info.get('files', {}).get('files', {})
```

Comment:
Typo?
```suggestion
        files = info.get('files', {})
```

Reply:
No the files entry contains a key called files which is what we want.

Reply:
Oh. That sounds hard to parse. Is it in the accepted standard already?

Reply:
It's just a byproduct of how I store the `FILES.json` object in info. I'll change the first key so they actually differ and remove the confusion.